### ISSUE TITLE
Add a `join` method

### ISSUE THREAD

**Issue by zhiburt** [zhiburt/tabled#31]
```rust
let table3 = table1.join(table2);
```

### Notes:

* ? Add an optional argument to define which stylies must be used
*  ? Add options to determine how to act if one table have a different amount of rows then the other

**Comment by senk8** [comment]
Hi @zhiburt , I'm interested in this.
Can I work on this?

**Comment by zhiburt** [comment]
Hi @senk8 

You're welcome.
Feel free to contact me if there was any questions.

I haven't noticed your comment in a first place,
I hope merge of #35 doesn't affected you :disappointed_relieved:

**Comment by zhiburt** [comment]
I can imagine the implementation could be something similar to https://github.com/zhiburt/tabled/blob/d5aa14dd2e0a2cdfe149f6ce9be874de0db1acff/src/disable.rs#L58-L75

Here we delete row.
But in `join` case we need to add rows/columns.

And just to note that `papegrid::Grid` is intentionally immutable from perspective of rows/columns. We can't add any rows and columns dynamically. 
We can only change data and styles.

*There are a bunch of things that must be documented :smile:

**Comment by zhiburt** [comment]
Note2: 

The method could be exposed via an object which implements `TableOption`.
This way we could set a different options.

Something like this.

```rust
table1.with(Join::inner(table2))
```

Not sure which is the best.
Just thoughts.

**Comment by zhiburt** [comment]
Note3:

I guess we shouldn't be worried about styles.
As we do in `Dissable` and `Panel`.

The style could be changed after `join`.

So essentially the main thing seems to copy the data.

**Comment by zhiburt** [comment]
Hey @senk8

Might you need some help?

**Comment by senk8** [comment]
Thank you, @zhiburt!

```rust
let table3 = table1.join(table2);
```

I tried to join the two grids together, using `Grid::get_content` to copy only the data, but this way the style information of `table1` and `table2` is lost.  How should I determine the style by default?

**Comment by senk8** [comment]
FYI 
https://github.com/senk8/tabled/blob/add-join-method/src/lib.rs#L266-L289

**Comment by zhiburt** [comment]
> FYI
> https://github.com/senk8/tabled/blob/add-join-method/src/lib.rs#L266-L289

Looks good.

> I tried to join the two grids together, using Grid::get_content to copy only the data, but this way the style information of table1 and table2 is lost. In this case, how should the style be determined ?

Actually never was thinking that way.
I was thinking that we will use all styles.

If you change your inner loop to the following.
It will work as I was thinking.

```rust
let cell_settings = if row < self.grid.count_rows() {
   self.grid.get_settings(row, column)
} else {
   let row = row - self.grid.count_rows();
   other.grid.get_settings(row, column)
};

new_grid.set(&Entity::Cell(row,column),cell_settings.border_restriction(false));         
```

This is join in `basic.rs`

```
┌─────────┬─────────────┬───────────┬─────────┐
│  name   │  based_on   │ is_active │ is_cool │
├─────────┼─────────────┼───────────┼─────────┤
│ Manjaro │ Arch        │ true      │ true    │
├─────────┼─────────────┼───────────┼─────────┤
│ Debian  │ Independent │ true      │ true    │
├─────────┼─────────────┼───────────┼─────────┤
│ Debian  │ Independent │ true      │ true    │
└---------+-------------+-----------+---------┘
  Manjaro | Arch        | true      | true     
  Debian  | Independent | true      | true     
```

**Comment by zhiburt** [comment]
Hhhm looking at this table I started to ask myself if `join` shall work like this.

I guess we need to merge columns not rows?

https://en.wikipedia.org/wiki/Join_(SQL)

Here's an example.

```rust
#[derive(Tabled)]
struct Person {
    name: String,
    age: usize,
}

#[derive(Tabled)]
struct Car {
    model: String,
    color: String,
    owner: String,
}

let people: Vec<Person> = ...
let cars: Vec<Car> = ....

let table = Table::new(people)
                          .with(Join::inner(Table::cars));

// result
   name  │ age │ model │ color │owner │
=======================================
  senk8    *   Ferari    red    senk8
```

Does this make sense?

Thoughts: `tuple`s already do such type of thing. So for achiving an example above `Vec<(Car, People)>` can be created

So what do you think about it?

**Comment by zhiburt** [comment]
But this interface is not very versatile ....
It would be cool if we could join tables by some rule e.g.

But that's fine for now.
Just thoughts.

**Comment by senk8** [comment]
Thanks for your rapid response !

> If you change your inner loop to the following.
> It will work as I was thinking.
> let cell_settings = if row < self.grid.count_rows() {
>    self.grid.get_settings(row, column)
> } else {
>    let row = row - self.grid.count_rows();
>    other.grid.get_settings(row, column)
>       new_grid.set(&Entity::Cell(row,column),cell_settings.border_restriction(false));  
> };

LGTM. But, If I use that way, it looks like this. For example, let table1 and table2 be as follows:

```
    let data1 = vec![
        (Developer("Terri Kshlerin"), Domain::Embeded),
        (Developer("Catalina Dicki"), Domain::Security),
        (Developer("Jennie Schmeler"), Domain::Frontend),
        (Developer("Maxim Zhiburt"), Domain::Unknown),
    ];

    let table1 = Table::new(data1).with(Style::psql());

    let data2 = vec![
        (Developer("Kshlerin"), Domain::Embeded),
        (Developer("Dicki"), Domain::Security),
        (Developer("Zhiburt"), Domain::Unknown),
        (Developer("Schmeler"), Domain::Frontend),
    ];

    let table2 = Table::new(data2).with(Style::psql());
    let table3 = table1.join(table2);
```

`table3` looks like the following.

```
      name       | Security | Embeded | Frontend | Unknown 
-----------------+----------+---------+----------+---------
 Terri Kshlerin  |          |    +    |          |         
 Catalina Dicki  |    +     |         |          |         
 Jennie Schmeler |          |         |    +     |         
  Maxim Zhiburt  |          |         |          |    +    
      name       | Security | Embeded | Frontend | Unknown 
-----------------+----------+---------+----------+---------
    Kshlerin     |          |    +    |          |         
      Dicki      |    +     |         |          |         
     Zhiburt     |          |         |          |    +    
```

I wish there was a way to just remove the `header`.  Do you know how to do this ?

`let other = other.with(Disable::Row(. .1));` because I'm trying to remove the header of `other`. However, I have confirmed that the header is not removed correctly this way. like this

```
      name       | Security | Embeded | Frontend | Unknown 
-----------------+----------+---------+----------+---------
 Terri Kshlerin  |          |    +    |          |         
 Catalina Dicki  |    +     |         |          |         
 Jennie Schmeler |          |         |    +     |         
  Maxim Zhiburt  |          |         |          |    +    
-----------------+----------+---------+----------+---------
    Kshlerin     |          |    +    |          |         
      Dicki      |    +     |         |          |         
     Zhiburt     |          |         |          |    +    
```

> Making this not a function but a component like other's could be cool? (a struct which implements TableOption, see all 
> others like Disable)

I think so ! I'm working on this.

**Comment by senk8** [comment]
> I guess we need to handle such case if self.grid.count_columns() != other.grid.count_columns(), probably by making 
> empty cells?

> Hhhm looking at this table I started to ask myself if join shall work like this.
> 
> I guess we need to merge columns not rows?
> 
> https://en.wikipedia.org/wiki/Join_(SQL)
> 
> Here's an example.

Oh! I was only thinking about the following vertical `join` case.

```
      name       | Security | Embeded | Frontend | Unknown 
-----------------+----------+---------+----------+---------
 Terri Kshlerin  |          |    +    |          |         
 Catalina Dicki  |    +     |         |          |         
 Jennie Schmeler |          |         |    +     |         
  Maxim Zhiburt  |          |         |          |    +    
-----------------+----------+---------+----------+---------
    Kshlerin     |          |    +    |          |         
      Dicki      |    +     |         |          |         
     Zhiburt     |          |         |          |    +    

```
I will be able to specify a horizontal `join`!

**Comment by senk8** [comment]
@zhiburt 

> I guess we need to handle such case if self.grid.count_columns() != other.grid.count_columns(), probably by making 
> empty cells?

> Hhhm looking at this table I started to ask myself if join shall work like this.
> 
> I guess we need to merge columns not rows?
> 
> https://en.wikipedia.org/wiki/Join_(SQL)
> 
> Here's an example.

Oh! I was only thinking about the following vertical `join` case.

```
      name       | Security | Embeded | Frontend | Unknown 
-----------------+----------+---------+----------+---------
 Terri Kshlerin  |          |    +    |          |         
 Catalina Dicki  |    +     |         |          |         
 Jennie Schmeler |          |         |    +     |         
  Maxim Zhiburt  |          |         |          |    +    
-----------------+----------+---------+----------+---------
    Kshlerin     |          |    +    |          |         
      Dicki      |    +     |         |          |         
     Zhiburt     |          |         |          |    +    

```

**Comment by zhiburt** [comment]
> let other = other.with(Disable::Row(. .1)); because I'm trying to remove the header of other. However, I have confirmed that the header is not removed correctly this way. like this
>
> 
> ```
>      name       | Security | Embeded | Frontend | Unknown 
>-----------------+----------+---------+----------+---------
> Terri Kshlerin  |          |    +    |          |         
> Catalina Dicki  |    +     |         |          |         
> Jennie Schmeler |          |         |    +     |         
>  Maxim Zhiburt  |          |         |          |    +    
>-----------------+----------+---------+----------+---------
>    Kshlerin     |          |    +    |          |         
>      Dicki      |    +     |         |          |         
>     Zhiburt     |          |         |          |    +    
>
>````

Your concern is this `-----------------+----------+---------+----------+---------` line?
Don't worry.
At least in my understanding its fine, we can't keep styles correct in any way.
Because its not clear what is correct :smile: 

Person then will able to call `Style` one more time and everything will be fixed.

```rust
let table1 = Table::new(data1).with(Style::pseudol());
let table2 = Table::new(data2).with(Style::psql());
let table3 = table1.join(table2).with(Style::Ascii);
```

**Comment by zhiburt** [comment]
> I have a question.
> 
> ```rust
> let table = Table::new(people)
>                           .with(Join::inner(cars));
> ```
> What is the meaning of this inner? Is it different from a vertical or horizontal join?

The idea was that it is a type of join.
The name essentially is not as important.
So essentially it could be `Join::vertical(cars)` and `Join::horizontal()`.

**Comment by zhiburt** [comment]
```rust
pub enum Join {
    Vertical(Table),
    Horizontal(Table),
}
```

Seems good :100:
that's what I meant ultimately.

Though I haven't tested it.

You can open a PR and we can discuss the rest there.

**Comment by senk8** [comment]
Hi, @zhiburt.

I have opened PR. Check it out 😄 .

**Comment by zhiburt** [comment]
I was looking if any library already has a similar functionality.
And I found out that `pandas` in `python` does support it.

What's interesting there, is that they actually looking for the same columns on vertical join.
They match 2 data sets columns and if there's the same column. They don't create a new one.

Your PR #40 covers only a case where columns names in the same way.

Here's an example what I mean.
You can use https://www.online-python.com/ to run it.

```python
import pandas as pd

test_list = [['a','b','c'], ['AA','BB','CC']]

data1 = pd.DataFrame(test_list, columns=['col_A', 'col_B', 'col_C'])
data2 = pd.DataFrame(test_list, columns=['col_A', 'col_2', 'col_3'])

# vertical
air_quality = pd.concat([data1, data2], axis=0)

print(air_quality.head())
```

Overall I think is a good option.
But then I tend to think it would be more valuable if we have a type information at match stage.

So I am not sure if your PR must include it as I even not sure if it's valuable.

Well free to express any ideas in case you have.

**Comment by senk8** [comment]
> But then I tend to think it would be more valuable if we have a type information at match stage.

I'm sorry, but what do you mean?  Is the "match" meant to be a "`match` expression"?

I think this is a very good idea 👍 . For some users, it may not be intuitive to have columns with the same name. There will be more options, and we will need to consider the design 🤔 .

**Comment by zhiburt** [comment]
> >   But then I tend to think it would be more valuable if we have a type information at match stage.
>
> I'm sorry, but what do you mean? Is the "match" meant to be a "match expression"?

I meant to say that when we do `table1.with(join::Join::Vertical(table2));` we don't know the real type of columns (rust type).
If we would have a `Table::join(data1, data2)` we would have the types.
So what I essentially meant is we could join data but not tables.

But overall I still don't see a use case for it.

Just thoughts.

**Comment by zhiburt** [comment]
The logic was discussed here was closed in #40

I have renamed @senk8's `Join` to `Concat` which more ergonomic name for this functionality.

The `join` logic its self must be determined yet.
As it turns out not to be as straightforward.
I'll open a new issue for this when the requirements will be more clear.
